### PR TITLE
Refactoring to version 9

### PR DESCRIPTION
**Pull request by amilosmanli**
Refactored code to firebase v9 modular

- Fixes issues with total item count in lazy mode

This is probably a breaking change

### CODE REVIEW

**Comment by amilosmanli** [comment]
@benwinding took me some time to clean up the branch. 

Most of the changes are in `yarn.lock` file, main code changes aren't that big

**Comment by matthieup240** [comment]
Hi, it will be merged soon ?

File: package.json [hunk 1]
```diff
@@ -22,12 +22,13 @@
...
     "@types/react-router-dom": "^5.3.3",
     "@types/rx": "^4.1.1",
-    "firebase": "^9.8.1",
+    "firebase": "^9.15.0",
     "firebase-tools": "11.x",
     "gulp": "^4.0.2",
     "jest": "^23.6.0",
     "microbundle": "^0.15.0",
     "prettier": "^2.8.3",
+    "prettier-plugin-organize-imports": "^3.2.2",
```
Comment (benwinding) [inline package.json:31]: This prettier package is an aesthetic choice, it shouldn't be included in this PR, as there's tonnes of import changes that have no effect on the behaviour of the system.

Mixing linting & functional changes in the same PR's makes the git history confusing and hard to see exactly what changes something.

I'll allow this, but next time, please restrict PR's to the smallest change necessary to make review easier. Linting PR's on the other hand, can be huge as they're _not functional changes_.

Comment (amilosmanli) [reply package.json:31]: @benwinding I reverted all the `prettier` and `tslint` changes, but this one slipped in somehow.

Let me know if you want me to revert this back.

Comment (benwinding) [reply package.json:31]: Yeah that would be great actually 🙏 Since this is a major firebase upgrade 👍

Comment (benwinding) [reply package.json:31]: Or just make another PR with the `prettier-plugin-organize-imports` linting changes, that can get merged in before (so these changes will be reduced)

**Comment by matthieup240** [comment]
I have a question about the collectionQuery with this new firebase version 9.

The usually `c.where('__name__', 'in', users)` syntax will not work.

How to do the equivalent ?

I found this in the Firestore doc :
`collectionQuery: c => query(c, where('__name__', 'in', users)`

This mean we have to import the query from the 'firebase/firestore' plugin ?

**Comment by amilosmanli** [comment]
@benwinding updated the PR. PTAL

**Review by benwinding — APPROVED** [review]
Thank you for this huge effort @amilosmanli !! 👏 
Much easier to see the changes, I'll merge this and deploy a newer minor release soon 👌

**Comment by benwinding** [comment]
Deployed as `react-admin-firebase@4.1.0` let me know how it goes! 🎉

**Comment by amilosmanli** [comment]
@benwinding thank you very much for accepting the changes! 

let me know if there are any bug reports for 4.1.0. Will be happy to help to resolve them

**Comment by ucefkh** [comment]
Hey @amilosmanli  I have this issue which normally should have been fixed by your changes! (and using getCountFromServer )

https://github.com/benwinding/react-admin-firebase/issues/272


Thanks for the hard work!

**Comment by CBrian04** [comment]
Hi everybody! 
Migrating collectionQuery from c.where() to query(c, where()) I got a firebase error saying
FirebaseError: [code=invalid-argument]: Type does not match the expected instance. Did you pass a reference from a different Firestore SDK?
Does someone had been confronted to that issue ? 

`{
                collectionQuery: (c: FireStoreCollectionRef) =>
                  query(c, where("status", "!=", "checkout"))
 }`

**Comment by MykolaRomaniv** [comment]
> Hi everybody! Migrating collectionQuery from c.where() to query(c, where()) I got a firebase error saying FirebaseError: [code=invalid-argument]: Type does not match the expected instance. Did you pass a reference from a different Firestore SDK? Does someone had been confronted to that issue ?
> 
> `filters={{ collectionQuery: (c: FireStoreCollectionRef) => query(c, where("status", "!=", "checkout")) }}`

Same error. Do you manage to fix it?

**Pull request merged**